floss/main.py

```diff
@@ -548,71 +441,111 @@ def load_workspace(sample_file_path, save_workspace):
                 "help (-h) for more information."
             )
         logger.info("Generating vivisect workspace...")
-    return viv_utils.getWorkspace(sample_file_path, should_save=save_workspace)
+    return viv_utils.getWorkspace(sample_file_path, should_save=False)
+
 
+class Architecture(str, Enum):
+    i386 = "i386"
+    amd64 = "amd64"
 
-def load_shellcode_workspace(sample_file_path, save_workspace, shellcode_ep_in, shellcode_base_in):
+
+def load_shellcode_workspace(
+    sample_file_path: str, shellcode_entry_point: int, shellcode_base: int, arch: Optional[Architecture] = None
+):
     if is_supported_file_type(sample_file_path):
-        logger.warning("Analyzing supported file type as shellcode. This will likely yield weaker analysis.")
+        logger.warning("analyzing supported file type as shellcode - this will likely yield weaker analysis.")
+
+    with open(sample_file_path, "rb") as f:
+        shellcode_data = f.read()
+
+    if not arch:
+        # choose arch with most functions, idea by Jay G.
+        candidates: List[Tuple[int, Architecture]] = []
+        for candidate in Architecture:
+            vw = viv_utils.getShellcodeWorkspace(
+                shellcode_data, candidate, base=shellcode_base, analyze=False, should_save=False
+            )
+            function_count = vw.getFunctions()
+            if function_count == 0:
+                continue
+
+            candidates.append((function_count, candidate))
 
-    shellcode_entry_point = 0
-    if shellcode_ep_in:
-        shellcode_entry_point = int(shellcode_ep_in, 0x10)
+        if not candidates:
+            raise ValueError("could not generate vivisect workspace")
 
-    shellcode_base = 0
-    if shellcode_base_in:
-        shellcode_base = int(shellcode_base_in, 0x10)
+        # pick the arch with the largest function count
+        (_, arch) = sorted(candidates, reverse=True)[0]
+
+        logger.info("detected shellcode arch: %s", arch)
 
     logger.info(
-        "Generating vivisect workspace for shellcode, base: 0x%x, entry point: 0x%x...",
+        "generating vivisect workspace for shellcode, arch: %s, base: 0x%x, entry point: 0x%x...",
+        arch,
         shellcode_base,
         shellcode_entry_point,
     )
-    with open(sample_file_path, "rb") as f:
-        shellcode_data = f.read()
-    return viv_utils.getShellcodeWorkspace(
-        shellcode_data, "i386", shellcode_base, shellcode_entry_point, save_workspace, sample_file_path
+
+    vw = viv_utils.getShellcodeWorkspace(
+        shellcode_data,
+        arch=arch,
+        base=shellcode_base,
+        entry_point=shellcode_entry_point,
+        should_save=False,
     )
 
+    vw.setMeta("StorageName", "%s.viv" % sample_file_path)
```

Comment:
what is this needed for? maybe add a comment in the code.

Reply:
i dont know, i just copied from capa :angel:

Reply:
https://github.com/fireeye/capa/commit/7665581c967a7eea630f05407e26377f3cdaf39d 😄

Reply:
Ah, this was part of https://github.com/fireeye/capa/pull/185 and ensures the workspace can be saved via `vw.saveWorkspace()`?!